### PR TITLE
add support for Cognito/Pinpoint integration via AnalyticsMetadata pa…

### PR DESCRIPTION
**Pull request by gspeicher**
This PR is a first-cut draft with minimal support to enable Cognito Analytics integration with Pinpoint.

For Cognito Analytics to work, it is not enough to enable it on the Cognito resource; you also need to include the AnalyticsMetadata parameter in Cognito API calls. Otherwise, email/phone endpoints will not be automatically created in Pinpoint for your users when they sign up. However, Amplify abstracts away all calls to the Cognito API, and it does not currently support the inclusion of that parameter, so Amplify developers are left with no simple way to communicate with their customers using Pinpoint.

Amplify already supports the creation and maintenance of Cognito and Pinpoint resources. In order to achieve full integration, it should support enabling Analytics on the user pool, as well as automatically including the metadata parameter in its Cognito API calls. This PR addresses the latter; it enables the use of the metadata parameter as a step towards the larger goal of fully automating the complete integration.

With this changeset in place, you can manually enable Cognito Analytics in the Cognito console under General Settings / Analytics, and add the following code snippet near the top of your main app source to configure the Auth service:

```
import {Auth} from 'aws-amplify';

Auth.configure({
  analyticsMetadata: {
    AnalyticsEndpointId: awsmobile.aws_mobile_analytics_app_id,
  },
});
```

Then your Cognito users will automatically have their Pinpoint endpoints created.

#### Description of changes
This change adds the analyticsMetadata config parameter to the Auth service so that it can be configured as shown above. That value then gets pushed into the `CognitoUserPool` provided by the `amazon-cognito-identity-js` package, and the value from the pool then gets used as the `AnalyticsMetadata` parameter in Cognito API requests where it is supported. Note that many of these requests happen from methods on `CognitoUser` so it simply gets the value through its existing `pool` attribute.

This is a minimal change with low risk for breakage and high impact in terms of opening up integration between Cognito and Pinpoint.

#### Issue #, if available
Issue #7675 

#### Description of how you validated changes
To validate these changes, I manually stepped through user signUp and email address verification and confirmed that the Pinpoint endpoint was created. More thorough testing and feedback from the community would be helpful and appreciated.

`yarn test` is failing for me with an `UnhandledPromiseRejection` in `@aws-amplify/core` but I believe that has something to do with my local environment and does not appear to be related to these changes. If anyone can verify this PR as non-breaking, that would be greatly appreciated.

I cannot think of any useful automated tests to add or change since this is an optional parameter that relies on a working environment to properly test. Perhaps we could validate the inclusion of the AnalyticsMetadata value in API requests when the value has been configured in Auth, but the architecture does not appear to lend itself to that test. If anyone can think of anything useful and feasible, please let me know. Otherwise the checklist item related to addition/modification of tests can likely be removed.

Also not sure the best place to add documentation for the steps necessary to enable and configure the integration as described above. Maybe the [Authentication - Getting Started](https://docs.amplify.aws/lib/auth/getting-started/q/platform/js) page? Thoughts and contributions welcome here as well.

#### Checklist

- [x] PR description included
- [ ] `yarn test` passes
- [ ] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by gspeicher** [comment]
Update: it seems that I (and others in online forums) misinterpreted the expected value for the `AnalyticsEndpointId` parameter in Cognito API calls. The documentation is rather sparse, but the expected value appears to be the ID of the actual endpoint that Cognito will create or update in Pinpoint. This poses somewhat of a chicken-and-egg problem, since the desired endpoint ID would need to be somehow known for a user prior to SignUp/SignIn. This could be as simple as using the user-supplied value in the username/login field when using password authentication, but it gets more complicated when using OAuth to login.

In addition, I'm not sure if this is a problem with my changes to the Auth library or with the underlying Cognito/Pinpoint integration, but when I create a user and verify their contact info (email or phone), Cognito ends up creating two distinct endpoints in Pinpoint: one gets created at the time of account creation (with the requested endpoint ID, OptOut:'ALL', and no Address), and the other gets created at the time of email/phone verification (generated endpoint ID, OptOut:'NONE', and expected Address). There is no obvious way to know the generated endpoint ID for the user so that the Auth service can be configured to include it in subsequent API calls. I have an active Technical Support case open with AWS Premium support to try and get to the bottom of what's going on here and see if there is a path forward.

Those extra endpoints with no valid Address are also causing downstream problems with Pinpoint Journeys; an issue for which I have a separate open support case and which I am told will be fixed in 3–4 weeks. In the meantime, the extra endpoints are preventing any messages from being delivered to the valid endpoints corresponding to the same user.

In conclusion, given the current state of affairs I would not recommend anyone plan to use this changeset in any production capacity until the behavior is better understood and any related underlying platform issues are resolved. For now, my solution is to create/maintain Pinpoint endpoints in the Cognito PostAuth trigger.
